Remove deprecated code paths

vyasr opened the pull request:
This PR removes outdated code that we no longer need to maintain. I've tried to keep the commits atomic, i.e. each commit corresponds to removing a related set of code. I have not yet removed the `flow/legacy.py` module because there are a couple of legacy support systems in place that I'm not sure about:

- I didn't think the `--full` argument to print_status was being removed, but it's currently only being taken care of by the compatibility layer.
- For `submit_operations`, at this point I'm a little confused about where exactly we've made `np` (and `ppn`) errors and where they are warnings; in `project.py` we error out in many of the cases. I think we can remove all the deprecated references, but want to confirm before doing it. Note that depending on what we decide I'll also need to remove those checks in `project.py`.
- JobsCursorWrapper may need to stay in for one more version, I think, since it fits our criteria for "core" API.

Comments on whether these changes look appropriate are welcome before I convert this from a draft PR. @mikemhenry @bdice I know you're less familiar with some of this old code, so feel free to skip reviewing if you'd like, but I added you to keep you in the loop!

vyasr commented on tests/test_project_legacy_api.py:
```diff
@@ -79,22 +74,6 @@ def submit_user(self, env, _id, operations, **kwargs):
-            raise unittest.SkipTest("Not implemented for Python 2.7.")
-        with self.assertWarns(DeprecationWarning):
-            return func(self)
-    return deprecation.fail_if_not_removed(wrapper)
```
I think I'm not understanding how this code works. According to those docs, the decorator raises an `AssertionError` if the test uses code that returns an `UnsupportedWarning`. However, the tests that I removed that were using the `fail_if_not_removed` decorator will fail with an `AttributeError` because there is no longer a `script` method defined for subclasses of `ComputeEnvironment`. Did you want to modify the `ComputeEnvironment` class (and any others with removed code) to raise these warnings when the relevant function is called, and then decorate all of the corresponding tests with `@unittest.expected_failure` so that we can confirm that the tests fail?

csadorf replied:
My general understanding is that you should increase the version number first and then use the unit tests to guide you in which functions *should* be removed.

So I just ran a local test, where I just checked out `develop` and increased the version number to `0.8.0` and I got a bunch of expected failures, but none of the unit tests raised any errors. This  might mean that our coverage is too low or that I misunderstand how this is supposed to work.

vyasr replied:
I don't think the issue is that our coverage is too low. I'm beginning to understand this a little better. You are correct that the expected use-case is that the relevant tests should raise warnings if deprecated APIs are still being used after version bumps. However, I think that the expected next step is that the tests would replace the deprecated code in the unit test with the newer API (and of course also delete the deprecated code paths in the code). In our case, I think we want to remove those deprecated tests entirely because they're superseded by the corresponding tests in the `test_project.py` API. Our tests already cover both APIs, and we check them directly rather than using them as part of more expansive, less "unit"-like tests, so I don't think the only result of trying to keep and fix the expected tests would be duplicating the tests in `test_project.py`. Correct me if you have a different understanding.

csadorf replied:
Yes, I think you are supposed to keep the tests, once any of them runs into an `AssertionError`, because of the version bump you remove corresponding code paths and either update or remove the tests.

I'm still a bit surprised that none of our legacy unit tests ran into any such assertion error on the `develop` branch.

vyasr commented on flow/environment.py:
```diff
@@ -259,15 +255,14 @@ class LSFEnvironment(ComputeEnvironment):
 
     Each compute node is assumed to have a specific number of compute units, e.g., CPUs.
     """
+    pass
```
You're right, this is not necessary. You can't have a naked class declaration, but docstrings are enough to make it valid syntax. I can remove this.

vyasr commented:
Regarding whether or not to deprecate anything new, I agree that this PR should be restricted to just removing existing deprecated code. Did I do (or say) something to suggest otherwise? If so, I will try and revert, because I agree that that will cause confusion.

csadorf commented:
> Regarding whether or not to deprecate anything new, I agree that this PR should be restricted to just removing existing deprecated code. Did I do (or say) something to suggest otherwise? If so, I will try and revert, because I agree that that will cause confusion.

No, this was more in response to a comment I made myself and then removed prior to submitting the review. I am sorry about the confusion that caused.

csadorf commented:
@vyasr Let me know when you want me to merge this!

vyasr commented:
Just to confirm, are we on the same page now with the tests that I removed, then?

I need to remove that extra `pass` that I added and add in support for the `full` argument into the main `print_status` API. Are we OK with just removing all references to `np`? If so, I can go ahead and do that as well. Once those tasks are complete it should be ready to merge.

csadorf commented:
> Just to confirm, are we on the same page now with the tests that I removed, then?
> 
> I need to remove that extra `pass` that I added and add in support for the `full` argument into the main `print_status` API. Are we OK with just removing all references to `np`? If so, I can go ahead and do that as well. Once those tasks are complete it should be ready to merge.

Yes, removed tests are fine. And OK on all the other things you mentioned. I am going to do one final review prior to merging so just go ahead with removing references to np at your discretion. Thank's!

vyasr commented:
@csadorf ready for review. The one non-deprecation that I snuck into this PR is documentation for the various `print_status` arguments; when I migrated the `--full` argument and added documentation it made sense to do the others as well.

csadorf commented on flow/project.py:
```diff
@@ -1041,13 +977,12 @@ def _print_progress(x):
                      unroll=True, compact=False, pretty=False,
                      file=None, err=None, ignore_errors=False,
-                     no_parallelize=False):
+                     no_parallelize=False, full=False):
```
Instead of introducing another (redundant) argument to the Python API, we should instead resolve the ``--full`` option as part of the argument parser handling. See [line 2622](https://github.com/glotzerlab/signac-flow/blob/98d64c83263864011b0939d1b1fc2ab6c9077739/flow/project.py#L2622) for an example.

csadorf commented on flow/project.py:
```diff
@@ -1079,11 +1014,59 @@ def print_status(self, jobs=None, overview=True, overview_max_lines=None,
+        :type all_ops:
+            bool
+        :param only_incomplete:
+            Remove all jobs that have no eligible operations.
```
"Only show jobs that have eligible operations."

The "remove jobs" is misleading.

csadorf commented on flow/project.py:
```diff
@@ -1079,11 +1014,59 @@ def print_status(self, jobs=None, overview=True, overview_max_lines=None,
+        :type ignore_errors:
+            bool
+        :param no_parallelize:
+            Don't parallelize the status update.
```
"Do not"

vyasr commented:
All set!
